clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,506 @@
+	networkTestRequestFilePath          = manifestPathPrefix + "/" + "networktestrequests.yaml"
+	crdManifestFilePath                 = manifestPathPrefix + "/" + "*CustomResourceDefinition.yaml"
+	clusterRoleBindingFilePath          = manifestPathPrefix + "/" + "roleBinding.yaml"
+	customResourceDefinitionName        = "networktestrequests.clusterloader.io"
+	rbacName                            = "networktestrequests-rbac"
+)
+
+var (
+	crdGvk = schema.GroupVersionKind{
+		Group:   "apiextensions.k8s.io",
+		Kind:    "CustomResourceDefinition",
+		Version: "v1"}
+
+	rbacGvk = schema.GroupVersionKind{
+		Group:   "rbac.authorization.k8s.io",
+		Kind:    "ClusterRoleBinding",
+		Version: "v1"}
+
+	crGvk = schema.GroupVersionKind{
+		Group:   "clusterloader.io",
+		Kind:    "NetworkTestRequest",
+		Version: "v1alpha1"}
+)
+
+func init() {
+	klog.Info("Registering Network Performance Measurement")
+	if err := measurement.Register(networkPerformanceMetricsName, createNetworkPerformanceMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", networkPerformanceMetricsName, err)
+	}
+}
+
+func createNetworkPerformanceMeasurement() measurement.Measurement {
+	return &networkPerformanceMeasurement{}
+}
+
+type networkPerformanceMeasurement struct {
+	k8sClient         kubernetes.Interface
+	dynamicClient     dynamic.Interface
+	resourceInterface dynamic.ResourceInterface
+	framework         *framework.Framework
+	numberOfServers   int
+	numberOfClients   int
+	podRatioType      string
+	testDuration      int
+	protocol          string
+	templateMapping   map[string]interface{}
+	// workerPodList stores list of podData for every worker node.
+	podInfo workerPodInfo
+
+	// metricVal stores MetricResponse received from worker node.
+	metricVal  [][]float64
+	metricLock sync.Mutex
+
+	podPairNumber int
+
+	stopCh chan struct{}
+	// startTimeStampForTestExecution is a futureTimeStamp sent to all pods to start
+	// sending the traffic at the same time.
+	startTimeStampForTestExecution int64
+
+	resultSummary testResultSummary
+}
+
+type podList []workerPodData
+type workerPodInfo struct {
+	podPairCh    chan uniquePodPair
+	workerPodMap map[string]podList
+}
+
+func (l *podList) Pop() (*workerPodData, error) {
+	if len(*l) == 0 {
+		return nil, fmt.Errorf("empty pod list")
+	}
+	lastElement := (*l)[len(*l)-1]
+	*l = (*l)[:len(*l)-1]
+	return &lastElement, nil
+}
+
+func (npm *networkPerformanceMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if err = npm.validate(config); err != nil {
+			return nil, err
+		}
+		if err = npm.start(config); err != nil {
+			return nil, err
+		}
+	case "gather":
+		summary, err := npm.gather()
+		if err != nil && !errors.IsMetricViolationError(err) {
+			return nil, err
+		}
+		klog.Info("metric: ", summary)
```

Comment:
This will be stored to file anyway - let's not log it.

Reply:
Done